fix(batcher): altda concurrent blobs respect strict holocene ordering rules

samlaf opened the pull request:
The recent holocene strict ordering rules broke the altda concurrent blob submission flow because concurrent requests to the da-server could be received out of order, and hence would be dropped by the op-node's derivation pipeline with strict ordering rules.

Solution TLDR: cache the altda commitments inside their respective channel, and add logic to pull them back in order.

This PR looks big but most of it is tests. The main Solution logic is quite small, and contained in this one commit: [3cbe31f](https://github.com/ethereum-optimism/optimism/pull/13169/commits/3cbe31fed3423229f5dbe17e29f1948de290488c)

It also contains a fix for the recently introduced sync_actions data race ([0a8bbe4](https://github.com/ethereum-optimism/optimism/pull/13169/commits/0a8bbe4f8b57cbbdf79f2db3ea1503dddf3ca60a)), but I think this will be subsumed by https://github.com/ethereum-optimism/optimism/pull/13330

Otherwise the added tests are:
1. e2e test that has the fakeDAServer return blobs out of order: [bebf7bb](https://github.com/ethereum-optimism/optimism/pull/13169/commits/bebf7bbe063d089a55f3cc28c139bff7e571941b)
2. unit tests for unhappy path failure cases: [a6c2830](https://github.com/ethereum-optimism/optimism/pull/13169/commits/a6c28304f3883d43e760640436588099790a95b3)
3. unit test that triggers channelTimeouts using the [failpoint](https://github.com/pingcap/failpoint) library: [a43ed4d](https://github.com/ethereum-optimism/optimism/pull/13169/commits/a43ed4df97d33a666a763da2095542556d668aef)
   - this one is more contentious as it requires some source code manipulation (b/c golang doesnt have macros) codegen. Updated the tooling (mise, just dependencies) to make it reproducible, and also using a separate GOCACHE for it, so should be very self-contained. But still left it as a separate commit so that we can easily remove it if you guys don't want it (even though I think you should, and should probably use it to test channelTimeouts for non-altda codepaths!)

samlaf committed:
test(e2e): new altda e2e test for concurrent blob submissions to maintain new holocene strict ordering rules

samlaf committed:
test(batcher): new unit tests for unhappy failure cases (currently only for altda)

test(batcher): fix flaky driver tests + speed them up

test(batcher): robustify batcher driver altda tests

samlaf committed:
fix(batcher): altda concurrent blob responses are reordered to respect holocene strict ordering rules

samlaf committed:
fix(batcher): sync_actions data race

computeSyncActions function of main goroutine is racing with the tx processing goroutine because it calls the channel functions directly without passing by the channelManager (which holds the lock). Fixed by passing channelManager lock to computeSyncActions

samlaf committed:
test(batcher): add failpoint-based test to trigger and test channelTimeout logic

scharissis commented:
Hello @samlaf and thank you for taking the time to make this contribution and suggestion to integrate failpoint to improve our testing suite.

I'm not too familiar with this particular codebase, so I am commenting more broadly. I am opposed to introducing failpoint, as the benefits do not seem to outweigh the costs. In particular:

- I dislike all the extra stuff which must be added/enabled for it
- It doesn't seem idiomatic
- I've also never seen it used before
- I've solved this kind of thing before with dependency injection; the test suite would implement a new implementation of the component you wish to modify (and/or test) and would pass that into the appropriate constructor

Having said that, I've raised this with the greater team to solicit more feedback around this.

samlaf commented:
> Hello @samlaf and thank you for taking the time to make this contribution and suggestion to integrate failpoint to improve our testing suite.
> 
> I'm not too familiar with this particular codebase, so I am commenting more broadly. I am opposed to introducing failpoint, as the benefits do not seem to outweigh the costs. In particular:
> 
> * I dislike all the extra stuff which must be added/enabled for it
> * It doesn't seem idiomatic
> * I've also never seen it used before
> * I've solved this kind of thing before with dependency injection; the test suite would implement a new implementation of the component you wish to modify (and/or test) and would pass that into the appropriate constructor
> 
> Having said that, I've raised this with the greater team to solicit more feedback around this.

That makes sense, I would probably also be against adding this if it was my codebase that I had to maintain. That being said, I ended up finding and using this for this test precisely becaue I couldn't find a way to get dependency injection to work. The place where it's injected is super deep inside a bunch of structs, and is automatically constructed by the batcher's channelManager. I really don't see a clean way to test this behavior without this (which may explain (?) why I don't think the current codebase has a test for this).

samlaf committed:
Revert "test(batcher): add failpoint-based test to trigger and test channelTimeout logic"

This reverts commit a43ed4df97d33a666a763da2095542556d668aef.

samlaf committed:
Revert "fix(batcher): sync_actions data race"

This reverts commit 0a8bbe4f8b57cbbdf79f2db3ea1503dddf3ca60a.
Reverting because this was fixed using another method, see https://github.com/ethereum-optimism/optimism/pull/13330.

samlaf committed:
Merge branch 'develop' into fix--batcher-altda-concurrent-blobs-respect-strict-holocene-order

samlaf committed:
test: fix batcher unit tests after merge

samlaf commented:
Reverted the failpoint-based test due to @scharissis's message above. Also reverted the data_race fix given that it's been fixed in another [PR](https://github.com/ethereum-optimism/optimism/pull/13330). Then merged master, fixed the conflicts, but now getting some errors when trying to run the e2e tests. See my discord message: https://discord.com/channels/1244729134312198194/1244729134848938031/1331788823285465128

samlaf committed:
test: fix TestBatcherCanHandleOutOfOrderDAServerResponses by using new DeployConfig

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates
